### ISSUE TITLE
Allow arguments to be passed to field default_factory callable

### ISSUE THREAD

**Issue by troyswanson** [jcrist/msgspec#339]
### Description

When defining a Struct, using the [field](https://jcristharif.com/msgspec/api.html#msgspec.field) function to define a default works great if you are passing a callable like [uuid4](https://docs.python.org/3.10/library/uuid.html#uuid.uuid4) which has no arguments.

However, if you need to provide the callable some arguments, there's some extra work to do.

An example is to create a new timezone-aware datetime object. The [datetime.utcnow](https://docs.python.org/3.10/library/datetime.html#datetime.datetime.utcnow) function returns the time in UTC, but it is timezone-naive (for reasons that make no sense to me, but whatever). The documentation in that function suggests that if you want a timezone-aware datetime object, use [datetime.now](https://docs.python.org/3.10/library/datetime.html#datetime.datetime.now) and pass [timezone.utc](https://docs.python.org/3.10/library/datetime.html#datetime.timezone.utc) as the argument.

A workaround that I've been able to use has been to create a "partial" object using the [functools.partial](https://docs.python.org/3.10/library/functools.html#functools.partial) function:

```python
from typing import Annotated
from uuid import UUID, uuid4
from datetime import datetime, timezone
from functools import partial
from msgspec import Struct, Meta, field

utcnow = partial(datetime.now, tz=timezone.utc)

class Namespace(Struct):
    namespace_name: str
    namespace_id: UUID = field(default_factory=uuid4)
    created_at: Annotated[datetime, Meta(tz=True)] = field(default_factory=utcnow)
```

This works pretty well, and I could see this being the recommended method. But right now it's not very clear from the documentation how this could be accomplished.

Happy to discuss further! Thanks for the package ❤️

**Comment by jcrist** [comment]
Hi! Thanks for opening this. Yeah, a `partial` or a `lambda` would be the recommended way to handle this.

Is there somewhere in the docs that you think this should be documented (and if so, would you be open to making a PR improving the docs here)?

**Comment by troyswanson** [comment]
Sure, I could take a swing at that.

Thanks for the response!

**Comment by jcrist** [comment]
Great! You might find the [contributing](https://github.com/jcrist/msgspec/blob/main/.github/CONTRIBUTING.md) page useful for getting started.

**Comment by jcrist** [comment]
I pushed a quick docs fix up in #347. If you don't find that sufficient, please feel free to push any additions as you see fit. Thanks!

**Comment by troyswanson** [comment]
That looks perfect, thank you!

Sorry for not contributing this myself. I owe you a coffee!
